### PR TITLE
Add OIDC SDK for Java

### PR DESCRIPTION
**Pull request by ChamathNS**
## Purpose
Add new module, **asgardio.io.java.oidc.sdk**

### CODE REVIEW

File: io.asgardio.java.oidc.sdk/pom.xml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  ~ under the License.
+  -->
+
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>java-oidc-sdk</artifactId>
+        <groupId>io.asgardio.java.oidc.sdk</groupId>
+        <version>0.1.2-SNAPSHOT</version>
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/pom.xml:26]: We need to start at 0.1.0

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        AuthenticationInfo authenticationInfo = new AuthenticationInfo();
+
+        try {
+            if (!requestResolver.isError() && requestResolver.isAuthorizationCodeResponse()) {
+                logger.log(Level.INFO, "Handling the OIDC Authorization response.");
+                boolean isAuthenticated = handleAuthentication(request, authenticationInfo);
+                if (isAuthenticated) {
+                    logger.log(Level.INFO, "Authentication successful. Redirecting to the target page.");
+                    return authenticationInfo;
+                } else {
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:114]: Do we actually need to have an else statement in the default flow?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+                    logger.log(Level.ERROR, "Authentication failed. Invalidating the session.");
+                    throw new SSOAgentServerException(
+                            SSOAgentConstants.ErrorMessages.AUTHENTICATION_FAILED.getMessage(),
+                            SSOAgentConstants.ErrorMessages.AUTHENTICATION_FAILED.getCode());
+                }
+
+            } else {
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:121]: Do we actually need to have an else statement in the default flow?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void logout(AuthenticationInfo authenticationInfo, HttpServletResponse response, String state)
+            throws SSOAgentException {
+
+        if (oidcAgentConfig.getPostLogoutRedirectURI() == null) {
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:138]: Empty URL strings allowed?
if not, better to use `StringUtils.isBlank()`

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        TokenRequest tokenRequest;
+        TokenResponse tokenResponse;
+
+        try {
+            authorizationResponse = AuthorizationResponse.parse(ServletUtils.createHTTPRequest(request));
+
+            if (!authorizationResponse.indicatesSuccess()) {
+                handleErrorAuthorizationResponse(authorizationResponse);
+                return false;
+            } else {
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:167]: Do we actually need to have an else statement in the default flow?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+                successResponse = authorizationResponse.toSuccessResponse();
+                authorizationCode = successResponse.getAuthorizationCode();
+            }
+            tokenRequest = getTokenRequest(authorizationCode);
+            tokenResponse = getTokenResponse(tokenRequest);
+
+            if (!tokenResponse.indicatesSuccess()) {
+                handleErrorTokenResponse(tokenRequest, tokenResponse);
+                return false;
+            } else {
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:177]: Do we actually need to have an else statement in the default flow?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        }
+    }
+
+    private void handleErrorTokenResponse(TokenRequest tokenRequest, TokenResponse tokenResponse) {
+
+        TokenErrorResponse errorResponse = tokenResponse.toErrorResponse();
+        JSONObject requestObject = requestToJson(tokenRequest);
+        JSONObject responseObject = errorResponse.toJSONObject();
+        logger.log(Level.INFO, "Request object for the error response: ", requestObject);
+        logger.log(Level.INFO, "Error response object: ", responseObject);
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:221]: Is the log level correct?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    private void handleErrorAuthorizationResponse(AuthorizationResponse authzResponse) {
+
+        AuthorizationErrorResponse errorResponse = authzResponse.toErrorResponse();
+        JSONObject responseObject = errorResponse.getErrorObject().toJSONObject();
+        logger.log(Level.INFO, "Error response object: ", responseObject);
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:228]: Is the log level correct?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    private TokenResponse getTokenResponse(TokenRequest tokenRequest) {
+
+        TokenResponse tokenResponse = null;
+        try {
+            tokenResponse = TokenResponse.parse(tokenRequest.toHTTPRequest().send());
+        } catch (com.nimbusds.oauth2.sdk.ParseException | IOException e) {
+            logger.log(Level.ERROR, "Error while parsing token response.", e);
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:237]: Can't we use the `handleErrorTokenResponse()` method here?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    private Map<String, Object> getUserAttributes(String idToken) throws SSOAgentServerException {
+
+        Map<String, Object> userClaimValueMap = new HashMap<>();
+        try {
+            JWTClaimsSet claimsSet = SignedJWT.parse(idToken).getJWTClaimsSet();
+            Map<String, Object> customClaimValueMap = claimsSet.getClaims();
+
+            for (String claim : customClaimValueMap.keySet()) {
+                if (!SSOAgentConstants.OIDC_METADATA_CLAIMS.contains(claim)) {
+                    userClaimValueMap.put(claim, customClaimValueMap.get(claim));
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:271]: If `customClaimValueMap.get(claim)` always returns a String can't we specify the data types of the map as `Map<String, String>`?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+    private void validateConfig(OIDCAgentConfig oidcAgentConfig) throws SSOAgentClientException {
+
+        validateForCode(oidcAgentConfig);
+    }
+
+    private void validateForCode(OIDCAgentConfig oidcAgentConfig) throws SSOAgentClientException {
+
+        Scope scope = oidcAgentConfig.getScope();
+        if (scope.isEmpty() || !scope.contains(SSOAgentConstants.OIDC_OPENID)) {
+            logger.error("scope defined incorrectly.");
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:290]: Is it necessary to both throw and log?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+            throw new SSOAgentClientException(SSOAgentConstants.ErrorMessages.AGENT_CONFIG_SCOPE.getMessage(),
+                    SSOAgentConstants.ErrorMessages.AGENT_CONFIG_SCOPE.getCode());
+        }
+
+        if (oidcAgentConfig.getConsumerKey() == null) {
+            logger.error("Consumer Key is null.");
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:296]: Is it necessary to both throw and log?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+            throw new SSOAgentClientException(SSOAgentConstants.ErrorMessages.AGENT_CONFIG_CLIENT_ID.getMessage(),
+                    SSOAgentConstants.ErrorMessages.AGENT_CONFIG_CLIENT_ID.getCode());
+        }
+
+        if (StringUtils.isEmpty(oidcAgentConfig.getCallbackUrl().toString())) {
+            logger.error("Callback URL is null.");
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:302]: Is it necessary to both throw and log?

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.Set;
+
+/**
+ * This class holds the constants used in the module, io.asgardio.java.oidc.sdk.
+ */
+public class SSOAgentConstants {
+
+    public static final String CONFIG_BEAN_NAME = "io.asgardio.java.oidc.sdk.config.model.OIDCAgentConfig";
+
+    // Oauth response parameters and session attributes
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java:32]: All comments should end with a `.`

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    public static final String ACCESS_TOKEN = "access_token";
+    public static final String ID_TOKEN = "id_token";
+    public static final String SESSION_STATE = "session_state";
+    public static final String USER = "user";
+
+    // Keystore file properties
+    public static final String KEYSTORE_NAME = "keystorename";
+    public static final String KEYSTORE_PASSWORD = "keystorepassword";
+
+    // application specific request parameters and session attributes
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java:43]: Comments should follow Sentence case format

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    public static final String OIDC_JWKS_ENDPOINT = "jwksEndpoint";
+    public static final String POST_LOGOUT_REDIRECTION_URI = "postLogoutRedirectURI";
+    public static final String AUTHENTICATED = "authenticated";
+    public static final String AUTHENTICATION_INFO = "authenticationInfo";
+    public static final String OIDC_OPENID = "openid";
+
+    // request headers
+    public static final String REFERER = "referer";
+
+    //context params
```
Comment (gayashanbc) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java:66]: Have a space before the beginning of the sentence

File: io.asgardio.java.oidc.sdk/pom.xml (hunk 1 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/pom.xml]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/f7dbfb11bb044bec17f2276d5bff60e622f0b2c0.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java (hunk 5 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/87be1ad8b44e103c81d6ce8a6ed22d4f6fc4f854.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java (hunk 7 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:212]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/87be1ad8b44e103c81d6ce8a6ed22d4f6fc4f854.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java (hunk 11 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/87be1ad8b44e103c81d6ce8a6ed22d4f6fc4f854.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java (hunk 12 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/87be1ad8b44e103c81d6ce8a6ed22d4f6fc4f854.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java (hunk 13 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/87be1ad8b44e103c81d6ce8a6ed22d4f6fc4f854.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java (hunk 14 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/87be1ad8b44e103c81d6ce8a6ed22d4f6fc4f854.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java (hunk 15 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/87be1ad8b44e103c81d6ce8a6ed22d4f6fc4f854.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java (hunk 16 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/SSOAgentConstants.java]: Resolved with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/87be1ad8b44e103c81d6ce8a6ed22d4f6fc4f854.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java (hunk 9 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:228]: I think moving the `handleErrorTokenResponse()` here would affect the readability of the code.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java (hunk 10 above)
Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:262]: Only the keys will be String format. Values can have different types (JSONArray, Arraylist, Date, etc.)

File: io.asgardio.java.oidc.sdk/pom.xml [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        <artifactId>java-oidc-sdk</artifactId>
+        <groupId>io.asgardio.java.oidc.sdk</groupId>
+        <version>0.1.0</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>io.asgardio.java.oidc.sdk</artifactId>
+    <packaging>jar</packaging>
+    <name>Asgardio - Java OIDC SDK core</name>
```
Comment (darshanasbg) [inline io.asgardio.java.oidc.sdk/pom.xml:33]: Shall we remove the "core" term here?

Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/pom.xml]: Fixed with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/af51f57c1a570de08c010334cfe63f93f6d70206.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void sendForLogin(HttpServletRequest request, HttpServletResponse response, String state)
+            throws SSOAgentException {
+
+        OIDCRequestBuilder requestBuilder = new OIDCRequestBuilder(oidcAgentConfig);
+        String authorizationRequest = requestBuilder.buildAuthorizationRequest(state);
+        try {
```
Comment (darshanasbg) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:90]: Add a new line.
https://www.oracle.com/java/technologies/javase/codeconventions-whitespace.html

Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/OIDCManagerImpl.java:91]: Fixed with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/1/commits/80e97e519132fde1a61fd5c9b4c71b09c7e7c426.

File: io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/bean/AuthenticationInfo.java [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+
+        this.user = user;
+    }
+
+    /**
+     * Returns the access token.
+     *
+     * @return The {@link AccessToken}.
+     */
+    public AccessToken getAccessToken() {
```
Comment (darshanasbg) [inline io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/bean/AuthenticationInfo.java:64]: Better to not to expose nimbus classes from the SDK APIs.

Comment (ChamathNS) [reply io.asgardio.java.oidc.sdk/src/main/java/io/asgardio/java/oidc/sdk/bean/AuthenticationInfo.java:64]: Fixed with https://github.com/asgardio/asgardio-java-oidc-sdk/pull/12/commits/999e40f1cc57d0a5b510d4c734790e448705b777.

**Review by darshanasbg — APPROVED** [review]

**Pull request merged**